render/render.js

```diff
@@ -628,65 +607,42 @@ module.exports = function($window) {
 			}
 		}
 		checkState(vnode, original)
-
+		var generation
 		// If we can, try to fast-path it and avoid all the overhead of awaiting
 		if (!mask) {
 			onremove(vnode)
-			removeChild(parent, vnode)
+			removeDOM(parent, vnode, undefined)
 		} else {
-			if (stateResult != null) {
-				var next = function () {
+			generation = currentRender
+			for (var dom of domFor(vnode)) delayedRemoval.set(dom, generation)
+			var finalizer = function(a, b) {
+				return function () {
 					// eslint-disable-next-line no-bitwise
-					if (mask & 1) { mask &= 2; if (!mask) reallyRemove() }
+					if (mask & a) { mask &= b; if (!mask) {
+						checkState(vnode, original)
+						onremove(vnode)
+						removeDOM(parent, vnode, generation)
+					} }
 				}
-				stateResult.then(next, next)
+			}
+			if (stateResult != null) {
+				stateResult.finally(finalizer(1, 2))
 			}
 			if (attrsResult != null) {
-				var next = function () {
-					// eslint-disable-next-line no-bitwise
-					if (mask & 2) { mask &= 1; if (!mask) reallyRemove() }
-				}
-				attrsResult.then(next, next)
+				attrsResult.finally(finalizer(2, 1))
```

Comment:
Can you revert the logic to what it was before (so we can keep the history)? Or does this shrink the bundle?

Reply:
Re. bundle size, I'll double check.

Reply:
Duplication is shorter after compression, thanks for the catch!